### ISSUE TITLE
Add @ExpectMaxDelete

### ISSUE THREAD

**Issue by jeanbisutti** [quick-perf/quickperf#69]
**Description**

The aim of this annotation is to check the maximum number of DELETE statements.

For example, a test method having `@ExpectMaxDelete(2)` annotation will fail if more than two DELETE statements are executed.

**Implementation steps**

1) Create the `@ExpectMaxDelete` annotation [here](https://github.com/quick-perf/quickperf/tree/master/sql/sql-annotations/src/main/java/org/quickperf/sql/annotation)

    Example: [@ExpectDelete](https://github.com/quick-perf/quickperf/blob/master/sql/sql-annotations/src/main/java/org/quickperf/sql/annotation/ExpectDelete.java)

2) Create a test class in [this package](https://github.com/quick-perf/quickperf/tree/master/sql/sql-integration-test/src/test/java) and add automatic tests

    Example: [ExpectDeleteTest.java](https://github.com/quick-perf/quickperf/blob/master/sql/sql-integration-test/src/test/java/ExpectDeleteTest.java)

3) In this [package](https://github.com/quick-perf/quickperf/tree/master/sql/sql-annotations/src/main/java/org/quickperf/sql/insert) add a performance measure extractor and a performance issue verifier. 

    Examples: 
    * [DeleteCountMeasureExtractor.java](https://github.com/quick-perf/quickperf/blob/master/sql/sql-annotations/src/main/java/org/quickperf/sql/delete/DeleteCountMeasureExtractor.java)
    * [NumberOfSqlDeletePerfIssueVerifier.java](https://github.com/quick-perf/quickperf/blob/master/sql/sql-annotations/src/main/java/org/quickperf/sql/delete/NumberOfSqlDeletePerfIssueVerifier.java)  
     * [MaxOfSelectsPerfIssueVerifier.java](https://github.com/quick-perf/quickperf/blob/master/sql/sql-annotations/src/main/java/org/quickperf/sql/select/MaxOfSelectsPerfIssueVerifier.java)

4) Create an AnnotationConfig in [SqlAnnotationsConfig](https://github.com/quick-perf/quickperf/blob/master/sql/sql-annotations/src/main/java/org/quickperf/sql/config/library/SqlAnnotationsConfigs.java). This new AnnotationConfig will use the previously created performance measure extractor, performance issue verifier together with a PersistenceSqlRecorder. 

    Example:
    ```java
        static final AnnotationConfig MAX_SQL_SELECT = new AnnotationConfig.Builder()
            .perfRecorderClass(PersistenceSqlRecorder.class)
            .perfMeasureExtractor(SelectCountMeasureExtractor.INSTANCE)
            .perfIssueVerifier(MaxOfSelectsPerfIssueVerifier.INSTANCE)
            .build(ExpectMaxSelect.class);
     ```

5) Use the new AnnotationConfig in `loadAnnotationConfigs()` method of [SqlConfigLoader](https://github.com/quick-perf/quickperf/blob/master/sql/sql-annotations/src/main/java/org/quickperf/sql/config/library/SqlConfigLoader.java) 

6) Complete the measure extractor and the performance issue verifier to make the tests green. Refactor to clean the code.

7) In [SqlAnnotationBuilder](https://github.com/quick-perf/quickperf/blob/master/sql/sql-annotations/src/main/java/org/quickperf/sql/annotation/SqlAnnotationBuilder.java), add a static method to can build the new annotation. This method can be used to define the annotation with [a global scope](https://github.com/quick-perf/doc/wiki/QuickPerf#annotation-scopes).

If you start to work on this issue, please leave a comment _I start to work on this issue_, so we can assign it to you. Do not hesitate to ask for information about the issue.

**Comment by rimow** [comment]
Hello, I would like to give a try to this issue. thanks !

**Comment by jeanbisutti** [comment]
@rimow Yes, of course!

I assign this issue to you.

Don't hesitate to leave a comment if I can help you.

**Comment by jeanbisutti** [comment]
@rimow 
Are you still interested in working on this issue? Can I help you?  I can help during [hack-commit-push](https://mailchi.mp/f7809cbbe48b/hack-commit-push-join-the-2020-virtual-event-participez-ldition-virtuelle-2020) virtual event on June the 27th.

**Comment by rimow** [comment]
Hello, yes still interested, give me 2 more days plz. thanks

**Comment by jeanbisutti** [comment]
Hello,
No problem, take your time. I just wanted to know if you are still interested.

**Comment by navkumar258** [comment]
Hi, want to work on this issue.Is it still open or is someone working on it?

**Comment by jeanbisutti** [comment]
@navkumar258 I believe that @rimow wants to work on this issue. 
I am going to create new issues and I could help during [hack-commit-push](https://mailchi.mp/f7809cbbe48b/hack-commit-push-join-the-2020-virtual-event-participez-ldition-virtuelle-2020) virtual event on June the 27th.

**Comment by navkumar258** [comment]
@jeanbisutti Sure, no prob.I will look out for other good first issues... 😃

**Comment by jeanbisutti** [comment]
@navkumar258 

I created other good first issues:
* https://github.com/quick-perf/quickperf/labels/%3Athumbsup%3A%20good%20first%20issue
* https://github.com/quick-perf/quickperf-examples/issues/12

Maybe you would be interested in working on one of them.

**Comment by rimow** [comment]
Hello @jeanbisutti  can you please tell me if there is a way to print some logs to debug ?

**Comment by jeanbisutti** [comment]
Hi @rimow,

You can use the debugger to debug. No logger is available (you can use System.out).

In your QuickPerf clone, don't hesitate to rebase on master and push your code so that I could help you.

**Comment by jeanbisutti** [comment]
Hi @rimow,

Can I help you? 

In your QuickPerf clone, don't hesitate to rebase on master and push your code so that I could help you.

Are you still interested in working on this issue?

Kind regards,

Jean

**Comment by heanzyzabala** [comment]
Hi @jeanbisutti 

I would like to work on this issue instead if I may.

**Comment by jeanbisutti** [comment]
Hello @navkumar258 and @heanzyzabala,

I  had no news from @rimow for more than three weeks.

@navkumar258 asked for working on this issue on 25 June, @heanzyzabala asked three days ago.

@navkumar258 are you still interested in working on this issue? 

If not, maybe that @heanzyzabala would still be interested.

Kind regards,

Jean

**Comment by navkumar258** [comment]
@jeanbisutti Yes, am still interested...thanks :smile:

**Comment by jeanbisutti** [comment]
@navkumar258 Great! I assign the issue to you.

@heanzyzabala [Other good first issues](https://github.com/quick-perf/quickperf/issues?q=is%3Aissue+is%3Aopen+label%3A%22%3Athumbsup%3A+good+first+issue%22) may interest you.

**Comment by navkumar258** [comment]
@jeanbisutti Made the changes as per the implementation steps in the issue description above, but after running mvn clean install, getting the below ClassCastException in the ExpectMaxInsert and ExpectMaxDelete tests.

`[ERROR] Failures:
[ERROR]   ExpectMaxDeleteTest.should_fail_if_the_number_of_sql_statements_is_greater_than_the_number_expected_with_annotation_on_class:76 expected:<[1]> but was:<[0]>
[ERROR]   ExpectMaxDeleteTest.should_fail_if_the_number_of_sql_statements_is_greater_than_the_number_expected_with_annotation_on_method:42 expected:<[1]> but was:<[0]>
[ERROR]   ExpectMaxInsertTest.should_fail_if_the_number_of_sql_statements_is_greater_than_the_number_expected_with_annotation_on_class:95
Expecting:
 <"
Time: 1.76
There was 1 failure:
1) execute_one_insert_but_no_insert_expected(ExpectMaxInsertTest$AClassAnnotatedWithExpectMaxInsert)
java.lang.ClassCastException: class com.sun.proxy.$Proxy33 cannot be cast to class org.quickperf.sql.annotation.ExpectMaxDelete (com.sun.proxy.$Proxy33 and org.quickperf.sql.annotation.ExpectMaxDelete are in unnamed module of loader 'app') at org.quickperf.sql.delete.MaxOfDeletesPerfIssueVerifier.verifyPerfIssue(MaxOfDeletesPerfIssueVerifier.java:9)`

**Note:** In the implementation step #3, the performance measure extractor and issue verifier are advised to be created in the /insert directory, but in my fork i have added them to the /delete directory, hope this is not causing the issue. Please check for anything i might be missing, i have already checked it and compared with other annotations like ExpectMaxInsert and ExpectMaxSelect. thanks :smile: 

Fork is [here](https://github.com/navkumar258/quickperf)

**Comment by jeanbisutti** [comment]
Hello @navkumar258,

Thank you for your work.

> @jeanbisutti Made the changes as per the implementation steps in the issue description above, but after running mvn clean install, getting the below ClassCastException in the ExpectMaxInsert and ExpectMaxDelete tests. (Fork is here)

To fix this issue, in `SqlAnnotationsConfigs` class, replace 
```java
    static final AnnotationConfig MAX_SQL_DELETE = new AnnotationConfig.Builder()
            .perfRecorderClass(PersistenceSqlRecorder.class)
            .perfMeasureExtractor(DeleteCountMeasureExtractor.INSTANCE)
            .perfIssueVerifier(MaxOfDeletesPerfIssueVerifier.INSTANCE)
            .build(ExpectMaxInsert.class);
```
by
```java
    static final AnnotationConfig MAX_SQL_DELETE = new AnnotationConfig.Builder()
            .perfRecorderClass(PersistenceSqlRecorder.class)
            .perfMeasureExtractor(DeleteCountMeasureExtractor.INSTANCE)
            .perfIssueVerifier(MaxOfDeletesPerfIssueVerifier.INSTANCE)
            .build(ExpectMaxDelete.class);
```

Could you please rebase on master? Be careful: today the `sql-integration-test`  Maven module was renamed into `sql-memory-test`.

Please apply the following code modifications:
* don't commit the `targetallocation-results.txt` file
* don't modify the exiting imports in `SqlAnnotationsConfigs ` and `SqlConfigLoader `
* properly align the code lines in `ExpectMaxDeleteTest`

After that, please create a PR. :)

Kind regards,

Jean

**Comment by navkumar258** [comment]
@jeanbisutti it was such a silly mistake, my bad. Incorporated the changes as suggested by you and opened a PR, please review. thanks :smile:

**Comment by jeanbisutti** [comment]
@navkumar258 [I reviewed the code.](https://github.com/quick-perf/quickperf/pull/97#pullrequestreview-459586939)

Could you please update it?

**Comment by navkumar258** [comment]
@jeanbisutti Done, please check and update. thanks !

**Comment by jeanbisutti** [comment]
Thank you @navkumar258

**Comment by jeanbisutti** [comment]
Fixed with 8e27123d7b2b044e5bdf50c7e6d4cd1e97416d5a
